core/src/main/scala/org/typelevel/otel4s/trace/SpanContext.scala

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2022 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s.trace
+
+import scodec.bits.ByteVector
+
+trait SpanContext {
+
+  /** Returns the trace identifier associated with this [[SpanContext]] as
+    * 16-byte vector.
+    */
+  def traceId: ByteVector
+
+  /** Returns the trace identifier associated with this [[SpanContext]] as 32
+    * character lowercase hex String.
+    */
+  def traceIdHex: String
+
+  /** Returns the span identifier associated with this [[SpanContext]] as 8-byte
+    * vector.
+    */
+  def spanId: ByteVector
+
+  /** Returns the span identifier associated with this [[SpanContext]] as 16
+    * character lowercase hex String.
+    */
+  def spanIdHex: String
+
+  /** Returns the sampling strategy of this [[SpanContext]]. Indicates whether
+    * the span in this context is sampled.
+    */
+  def samplingDecision: SamplingDecision
+
+  /** Returns `true` if this [[SpanContext]] is valid.
+    */
+  def isValid: Boolean
+
+  /** Returns `true` if this [[SpanContext]] was propagated from a remote
+    * parent.
+    */
+  def isRemote: Boolean
+}
```

Comment:
I think it would be a good idea to make this sealed to provde "safe" and "unsafe" constructors for `SpanContext`s with `ByteVectors`. There should probably also be a constructor that safely generates a context from `Random`.

This might make `isValid` redundant of course. I suppose the question is do we want to pass through invalid contexts parsed/converted from upstream system and whether we want to differentiate between a no-op and invalid data. My hunch is that it's probably a good idea to represent a no-op as a `None` and invalid as `false` here, to aid debugging.

Reply:
I was thinking about a simple [wrapper](https://github.com/typelevel/otel4s/pull/37/files#diff-9a1644ea341d7f1a5682ebe47f65524c64b63ceaeb1280a8947bf26f0f9c3970) for java implementation.

Safe and unsafe constructors make sense to me. They are useful for third-party integrations (i.e. creating SpanContext from HTTP request headers). I will add them.